### ISSUE TITLE
Negative gradients when calculating GradCAM 

### ISSUE THREAD

**Issue by ashishsb0307** [jacobgil/pytorch-grad-cam#235]
I have a Segmentation network model trained for 2 classes and am able to see accurate results. But when using grad-cam for the heatmap, I am able to see good results for the last convolution layer for both the classes but having issues when trying to generate a heatmap for the second last convolution layer for one of the classes (the other class's heatmap is working fine).

```
**Last 5 layers**
convolution_layer(filters:8, kernel:3*3)
convolution_transpose_layer(filters:2, kernel:2*2)
convolution_layer(filters:2, kernel:3*3)
convolution_layer(filters:10, kernel:1*1)
activation_layer(softmax)
```
The heatmap is empty because of all negative pooled gradients(due to mean from all the -ve gradients wrt Conv layer), resulting in negative values in pooled_grads*convolution_output on which relu is applied, giving all zeros.

**What does it mean for GradCAM to be all negative?

Why is it that all channels in the convolution lead to a "negative" contribution to the true output class?**

`https://arxiv.org/pdf/2002.11434.pdf following this paper for heatmap for segmentation models.`

**Comment by jacobgil** [comment]
Then negative gradients mean that the activations of that layer pull the network away from that category.
In Resnet for example, between the activations of the last CNN layer and the classification, there is a fully connected layer, so it's a linear system.
In such a linear system, the gradient is very informative and tells us about how the activations affect the output.

However -
As you move away from the last layer, the system in between the target layer, and the model output, is more complex and less linear.
For example - in Unet which I think is the network you're using, you have more concatenations with early layers, and more non linear activation functions.

So this means that as you go back, GradCAM won't necessarily work great.

I suggest you try one of the gradient free methods as well and see if they also suffer from this: 
AblationCAM or ScoreCAM.
Does it look the same for them?

**Comment by ashishsb0307** [comment]
> 



> Unet which I think is the network you're using, you have more concatenations with early layers, and more non linear activation functions.

Yes, in fact, i am using small-Unet, but these concatenations with the earlier layer happen in Concat layer occurring 2 layers before the penultimate convolution layer(of which i am trying to get the heatmap wrt).

On model inferencing, I am able to see semantic segmentation for both classes in the same image, but HeatMap for one class is empty in this penultimate convolution layer.
